lib/libc/stdlib/malloc/mrs/mrs.c

```diff
@@ -1255,11 +1255,6 @@ static void *mrs_realloc(void *ptr, size_t size) {
 	size_t old_size = cheri_getlen(ptr);
 	mrs_debug_printf("mrs_realloc: called ptr %p ptr size %zu new size %zu\n", ptr, old_size, size);
 
-	if (size == 0) {
-		mrs_free(ptr);
-		return NULL;
-	}
-
 	if (ptr == NULL) {
```

Comment:
```
	if (ptr == NULL) {
 		return mrs_malloc(size);
 	}
	void *new_alloc = mrs_malloc(size);
	/* old object is not deallocated according to the spec */
	if (new_alloc == NULL) {
		return NULL;
	}
	memcpy(new_alloc, ptr, size < old_size ? size : old_size);
	mrs_free(ptr);
	return new_alloc;
```
can also just be
```
	void *new_alloc = mrs_malloc(size);
	/* old object is not deallocated according to the spec */
	if (ptr != NULL && new_alloc != NULL) {
		memcpy(new_alloc, ptr, size < old_size ? size : old_size);
		mrs_free(ptr);
	}
	return new_alloc;
```
(maybe with a tweaked comment)

Reply:
True, but probably as a separate followup after the size == 0 fix?